### ISSUE TITLE
Using Pubs to improve onboarding, troubleshooting, and overall experience

### ISSUE THREAD

**Issue by pfraze** [ssbc/phoenix#357]
There's a lot of different but related issues which I think we can improve in one swoop

 - Difficulty troubleshooting sync.
   - There's no way to know if you have a sync-path to a user.
   - It should be very obvious who a Pub is hosting, especially for new users. That's not currently given focus. (Related, #301.)
   - There's no way to measure the freshness of sync with a user, once you have a sync-path. "You get the message yet? How about now? Now? Wait, which message?"
 - Sub-par identity and name sharing.
   - Somebody entering the network may/almost-assuredly-will have trouble discerning identities thanks to dead feeds and, possibly, impersonators (see #327). Manual confirmation by sharing pubkey hashes or invite codes doesn't scale.
   - We need a good, clear system and UX for using the names assigned by other users. The "trust" button was unclear and a little buried, and for people to get oriented quickly, I think setting up names-assigners should be a first step.
 - No UI controls for pubs. 
  - You have to add pubs to your gossip table via invites or by manually publishing a `type: pub` on the CLI. 
  - There's no way to add a pub without broadcasting it.
  - There's no way to remove a pub at all. At some point, 192.168.1.61 got published by accident, and I've been attempting syncs to it ever since.
  - There's some concern with auto-connecting to pubs, see https://github.com/ssbc/scuttlebot/issues/68

I think we should view Pubs as our biggest organizing tool on the network.

 - They're vital for sync over the Internet, so they're almost always involved.
 - Getting on a Pub involves invite codes and direct connections, which we can leverage into a two-way identity confirmation. Pubs can also leverage domain/CA certs.
 - Pub-operators are savvy users who we can vest responsibility in.
 - Pub membership is exclusive, can be organizational, and should signal identity.
 - Pubs are HTTP accessible, and so they could make good user directories.

Pubs have a spiritual connection to PGP's names servers, but they already assume much more responsibility for SSB's infrastructure than names servers did for PGP.

So, here's what I recommend:

 - Implement a Pub frontend (maybe a different repo than Phoenix?) that shows who is hosted there, and gives info about the Pub itself and SSB.
 - Implement controls to add/remove your Pubs. Make this a prominent feature to a new user -- maybe include it during setup?
 - Continue publishing pub announcements, and use that to give recommendations to the Pub controls. However, don't automatically publish a pub announcement when you add a Pub -- maybe do that via a "share" button.
 - Prominently list on user profiles which Pubs they're members of.
 - Add an Address Book View to user profiles, to see their contact list. Include a "Use this address book" toggle in order to trust their names.
 - When an invite code is used, open that Pub's address book, and possibly highlight the "Use this address book" button if it's not already toggled on.
 - (This may be a privacy issue, but) Have the pubs report the last time they got a new message from a given user, and show that on the profiles. (This would help troubleshoot synchronization)

By the way, [.pub is now a valid gTLD](https://www.namecheap.com/domains/new-tlds/explore.aspx?keyword=pub), and that just makes me, so happy. I snagged ssbc.pub!

**Comment by pfraze** [comment]
Some joker decided to help illustrate our names situation

![screen shot 2015-03-30 at 12 08 22 pm](https://cloud.githubusercontent.com/assets/1270099/6901837/932ac602-d6d5-11e4-8a07-497e2326e100.png)

**Comment by pfraze** [comment]
Which reminds me -- you might think "sure, what are the odds of somebody attacking the network?" But that's you forgetting how programmers react to new technology. It comes in two steps:

 1. cool, what's it do?
 2. cool, can I break it?

**Comment by pfraze** [comment]
Also a nicely-timed HN thread about decentralized networks (Usenet) and the Web of Trust:

https://news.ycombinator.com/item?id=9291709

**Comment by pfraze** [comment]
[Advogato's Trust Metric](http://www.advogato.org/trust-metric.html)

**Comment by ahdinosaur** [comment]
yes :+1:

**Comment by jtremback** [comment]
How is diaspora doing? How will a major reliance on pubs affect the perception and adoption of the network? Will the developers of the software run the main hubs? If not, then who? Why?

**Comment by pfraze** [comment]
@jtremback this change would not introduce Pubs into the system for the first time. We're already reliant on them for synchronizing outside of firewalls/NATs. This proposal is mainly to add a UI for managing the Pubs, and to use them to improve identity confirmation.

I'm mindful of the Diaspora failure. For what it's worth, Diaspora was a host-based network, which SSB is not, even with the inclusion of pubs. Pubs are more similar to mailservers than to Diaspora's pods.

@dominictarr has been pretty adamant about pushing Pubs into the background. I understand why, it's always better when things just work. This proposal explains why I'm in favor of pulling Pubs into the UI more, but I'm open to arguments otherwise. There's a balance we can strike somewhere.

As for who runs the pubs, for now, it's us, but eventually we'll need to give a good incentive for other people to run them. We'll figure that out later.

**Comment by jtremback** [comment]
Definitely... I'm not trying to say this is a bad idea. Just be mindful of the fact that using pubs to patch shortcomings in the underlying network and basing a lot of your identity, functionality and network topography around them are very different things. Pubs should be a removable polyfill IMO. 

You've spent a lot more time working on this and thinking about it than me of course, but the idea of:

- Pub membership is exclusive, can be organizational, and should signal identity.

is one I particularly disagree with.

**Comment by pfraze** [comment]
What do you dislike about that specific point?

**Comment by dominictarr** [comment]
@pfraze I also have reservations about the point @jtremback mentions. But the main thing right now is that I don't know quite what it means... what does "exclusive" mean? do you mean "invite only"? I can be on many pubs.

I agree with @jtremback but I am not against having special roles, as long as the user has the freedom to choose who performs those roles for them, and that things are immutable and secure.
I'd be against having something like email addresses (name@host) (though if host was a key it could be acceptable)

but the problem with this issue @pfraze is that it's not entirely clear what you are proposing. Some of this sounds like simple edge case fixes (like to delete pubs) other things are less clear (@jtremback mentions)

**Comment by pfraze** [comment]
Exclusive is referring to invite-only. That you got an invite can be used as an identity signal, and (possibly!) a fun one. It would mean that you're friends with the pub owner, or part of the pub's org (like if we made a pub for ssbc) or just happened to be there when invites were getting handed out. The meaning would depend on how the pub is run.

By showing your pub memberships prominently, the idea is to

 1) make it easier to evaluate who someone is
 2) treat the pub-membership a bit like badges/achievements

So that's what "Pub membership is exclusive, can be organizational, and should signal identity" meant. I think "should" might have been a problematic word choice -- I meant should as in "is likely to."

**Comment by pfraze** [comment]
For deciding identity, the most valuable signal somebody can emit is "I checked this user's pubkey, and I'm confident I can identify them correctly." (Let's call this the "confirmation" signal.) This is what PGP's signatures communicate. It's a tedious signal to generate, because it involves manually sharing the pubkey/hash-id out of band. 

This proposal doesn't include the explicit confirmation signal; instead it suggests using Pub membership as an alternative signal. The Pub operator can choose to do confirmation during the invite process, in which case membership implies confirmation. (For instance, this would almost certainly be the case for an organization's Pubs.) However, invites don't require confirmation, so membership is a weaker signal than explicit confirmation is.

Why use membership, then, instead of explicit confirmation?

 1. **To avoid tedium.** People will be disinclined to confirm pubkeys frequently. Since Pubs are a necessary part of WAN sync, using membership as a signal is meant to kill two birds with one stone. If you're syncing over the LAN, and don't have any Pub involved, then you probably don't have an identity confirmation problem -- it's fairly painless to compare IDs in person -- so the lack of Pubs isn't an issue.
 2. **To avoid bad signals generated by unsavvy users.** If we have users do confirmations, we'll need to make sure they understand the tools and don't generate meaningless confirmations. There are lots of ways to do this, but Pub operators are a self-selecting crowd of savvy users, and so (in theory) they should generate more reliable signals.
 3. **To create a more intuitive signal.** If you don't understand the model behind async crypto, you might not understand what identity confirmation implies, or why it's significant. My thinking is that Pub membership could be more intuitive, because it's rooted in the idea of exclusivity. This would be most effective for Pubs owned by friends or organizations -- because, hey, if you're on Bob's pub, you're probably the Alice that Bob (and I) know!

Right solution? Don't know, but that's the logic.

**Comment by jtremback** [comment]
While pub servers may currently be necessary for the operation of the network, they are not good for the UX and 'gestalt' of the network. I mean at what point do you just set up a mysql db and make a conventional CRUD app with an RPC for communication between instances? (Diaspora) If everything relies on these central servers, doesn't all the fancy scuttlebutting start to seem besides the point?

As far as I can recall, the only 'hard' requirement of pub servers is for routing on connect. The rest of what they are used for is essentially optimization around uptime and convenience around naming. Even routing on connect will soon be doable in a decentralized manner: https://www.youtube.com/watch?v=oSnYmMHp-o8

I believe that the uptime issue can be solved by better rules around data replication by friends of friends, and possibly incentivization of replication as we have talked about previously.

I'm not against the use of pub servers, but I believe they should be viewed as a temporary polyfill for some of the undesirable characteristics of the network that we currently need to run on. I was against hardcoding IP's into the message format for similar reasons.

Anyway though, my concerns are all theoretical and I understand your desire to have a social network that works as well as possible, today. Carry on with the good work!

**Comment by pfraze** [comment]
@jtremback I understand your concern, it's not being ignored. I really have no interest in creating an experience or an architecture like Diaspora.

If you have proposals to deprecate Pubs, please post them as a separate issue! because a Publess network would be great. But this proposal is based on the assumption that Pubs are going to be used -- so let's make sure they work well, and perhaps eek more value out of them.

I don't want to spend more time on it here, but one point I will quickly address:

> As far as I can recall, the only 'hard' requirement of pub servers is for routing on connect.

That's not right. Pubs receive messages from the users they follow, and rehost the messages to others. They're more like mailmen than routers.

**Comment by dominictarr** [comment]
is pub membership a strong signal?

it's really easy to overshare an invite. (current) invites are multiuse, and can easily be tweeted or shared in chat etc. I would fully expect people to publicize invites to their severs, especially while they are getting started.

**Comment by ahdinosaur** [comment]
> is pub membership a strong signal?

the meaning of pub membership depends on the pub. for example, if [Enspiral](http://enspiral.com) were to have a pub, it would be mean only "contributors" ([people who are within the cell membrane of Enspiral](http://whatis.enspiral.com/#6)) are members of the pub, which signals something.

**Comment by pfraze** [comment]
@ahdinosaur that's my thinking.

It does make the signal "soft," you have to understand the pub to understand the signal.
To avoid that, we might make identity-confirmation a separate signal that users and pubs can create.

**Comment by dominictarr** [comment]
@pfraze @ahdinosaur We'd want a different invite mechanism if we where to realistically enforce that constraint. Like, a manual add, not an invite code. In this case, the "pub" is more like some sort of club or group.

**Comment by ahdinosaur** [comment]
@dominictarr an invite code still works if it can only be redeemable once, as it works well within the typical "please accept your invitation to <group>" email as we are all used to.

**Comment by pfraze** [comment]
Yeah that's how I use them -- I issue them one at a time as needed

**Comment by dominictarr** [comment]
hmm, using a narrowish expiry (like 24 hours) would make it fairly safe.
Okay so if you are doing it like that, pub membership is ~ the same as being following by a prominent user
